[FLINK-34111][table] Add support for json_quote, json_unquote, address PR feedback #24156

anupamaggarwal opened the pull request:
## What is the purpose of the change

Adds implementation of `json_quote` and `json_unquote` functions.
PR is a continuation of https://github.com/apache/flink/pull/24156 and addresses feedback received on the original PR.


## Brief change log / changes compared to PR 24156
#### Diff compared to https://github.com/apache/flink/pull/24156 - [link](https://github.com/anupamaggarwal/flink/compare/jeyhun-34111..34111-jsonquote_unquote)

- Change implementation for `json_quote`, now with escaping logic
- Change implementation for `json_unquote` to unescape special characters
- Delegates valid json checking to SqlJsonUtils#isJsonValue
- Add/ update some additional tests
- Updates documentation to remove reference to MySql utf8mb4 encoding


## Open questions
- High level behavior for invalid json in json_unquote (pass original input as is, Vs throwing exception) 
- To check for JSON_VALIDITY for `json_unquote` logic for `IS JSON` is used but in certain cases `IS JSON` behaviour does not seem to be consistent with Json spec as documented here https://www.json.org/json-en.html 
  - According to json spec `""""` should not be a valid json (which is also flagged by JsonLint) but `select '""""' IS JSON` returns true 
       -  In case of invalid json string for json_unquote an exception is thrown in case of Mysql (However we won't do the same) 


### References
- postgres implementation for [escape](https://github.com/postgres/postgres/blob/524d64ea8e3e49b4fda41ff9b2f048b697384058/src/backend/utils/adt/json.c#L2458) 

## Verifying this change

- Added ITs in JsonFunctionsITCase

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

anupamaggarwal commented:
@flinkbot run azure

anupamaggarwal commented:
@flinkbot  run azure

anupamaggarwal commented:
@flinkbot run azure

anupamaggarwal commented:
@flinkbot  run azure

snuyanzin commented:
Thanks for the contribution @anupamaggarwal 
I checked for most cases it works well

I found one issue with symbols encoded outside of [BMP ](https://en.wikipedia.org/wiki/Plane_(Unicode)#Basic_Multilingual_Plane) 

e.g.
```sql
SELECT '≠';
```
This works fine
now if we try it with newly added functions
```sql
SELECT json_quote('≠');
```

it fails as
```
Caused by: java.util.IllegalFormatConversionException: x != java.lang.Character
	at java.base/java.util.Formatter$FormatSpecifier.failConversion(Formatter.java:4426)
	at java.base/java.util.Formatter$FormatSpecifier.printInteger(Formatter.java:2938)
	at java.base/java.util.Formatter$FormatSpecifier.print(Formatter.java:2892)
	at java.base/java.util.Formatter.format(Formatter.java:2673)
	at java.base/java.util.Formatter.format(Formatter.java:2609)
	at java.base/java.lang.String.format(String.java:2897)
	at org.apache.flink.table.runtime.functions.scalar.JsonQuoteFunction.formatStr(JsonQuoteFunction.java:77)
	at org.apache.flink.table.runtime.functions.scalar.JsonQuoteFunction.quote(JsonQuoteFunction.java:68)
	at org.apache.flink.table.runtime.functions.scalar.JsonQuoteFunction.eval(JsonQuoteFunction.java:88)
	at StreamExecCalc$3.processElement(Unknown Source)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:75)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:50)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:29)
	at ...
```

I tend to think that here
```
	at org.apache.flink.table.runtime.functions.scalar.JsonQuoteFunction.formatStr(JsonQuoteFunction.java:77)
	at org.apache.flink.table.runtime.functions.scalar.JsonQuoteFunction.quote(JsonQuoteFunction.java:68)
	at org.apache.flink.table.runtime.functions.scalar.JsonQuoteFunction.eval(JsonQuoteFunction.java:88)
```
need to work with codepoints rather than chars


I also checked same query for MySQL and it works perfectly fine

anupamaggarwal commented:
thanks for your review @snuyanzin ! 
Fixed in [4267018](https://github.com/apache/flink/pull/24967/commits/4267018323dc77773bfa1d65ee9fcb49b024e03d)

snuyanzin commented on docs/data/sql_functions.yml:
```diff
@@ -377,6 +377,12 @@ string:
   - sql: SUBSTR(string, integer1[, integer2])
     table: STRING.substr(INTEGER1[, INTEGER2])
     description: Returns a substring of string starting from position integer1 with length integer2 (to the end by default).
+  - sql: JSON_QUOTE(string)
```
One thing I forgot to mention: also need to update chinese doc.
It's ok to put there english version of description however it's better to keep them synced

anupamaggarwal replied:
Created JIRA https://issues.apache.org/jira/browse/FLINK-35800 (and added a comment to original JIRA https://issues.apache.org/jira/browse/FLINK-34111)

snuyanzin replied:
Thanks for creation of jira
I would still vote for having filled chinese doc with english description which then could be translated within the jira issue you've created

snuyanzin reviewed (changes requested):
hm during testing I noticed 2 strange things
1. if escaped quote is alone or with something not UTF-8 bytes related then everything is ok e.g.
```sql
SELECT JSON_UNQUOTE('"\""'), JSON_UNQUOTE('"\"\t"');
```
it returns 
```
"                              "
```
which is ok, however once we add utf-8 bytes like
```sql
SELECT JSON_UNQUOTE('"\"\ufffa"');
```
it starts printing backslash as well which is NOT OK

2.It seems current implementation allows to have multiple double quotes
e.g. this query is still working (in MySQL it fails )
```sql
SELECT JSON_UNQUOTE('"1""1"');
```
From one side MySQL doesn't allow this from another side I think it is still ok if it returns valid result.
However the main issue is that it fails (trace below) if we use such approach for invalid utf e.g.
```sql
 SELECT JSON_UNQUOTE('"1""\uzzzz"');
```
at the same time this query works fine
```sql
SELECT JSON_UNQUOTE('"1\uzzzz"');
```

```
Caused by: org.apache.flink.util.FlinkRuntimeException: java.lang.NumberFormatException: For input string: "zzzz" under radix 16
	at org.apache.flink.table.runtime.functions.scalar.JsonUnquoteFunction.eval(JsonUnquoteFunction.java:118)
	at StreamExecCalc$3.processElement(Unknown Source)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:75)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:50)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:29)
	at org.apache.flink.streaming.api.operators.StreamSourceContexts$ManualWatermarkContext.processAndCollect(StreamSourceContexts.java:425)
	at org.apache.flink.streaming.api.operators.StreamSourceContexts$WatermarkContext.collect(StreamSourceContexts.java:520)
	at org.apache.flink.streaming.api.operators.StreamSourceContexts$SwitchingOnClose.collect(StreamSourceContexts.java:110)
	at org.apache.flink.streaming.api.functions.source.InputFormatSourceFunction.run(InputFormatSourceFunction.java:100)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:113)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:71)
	at org.apache.flink.streaming.runtime.tasks.SourceStreamTask$LegacySourceFunctionThread.run(SourceStreamTask.java:338)
Caused by: java.lang.NumberFormatException: For input string: "zzzz" under radix 16
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
	at java.base/java.lang.Integer.parseInt(Integer.java:668)
	at org.apache.flink.table.runtime.functions.scalar.JsonUnquoteFunction.fromUnicodeLiteral(JsonUnquoteFunction.java:99)
	at org.apache.flink.table.runtime.functions.scalar.JsonUnquoteFunction.unquote(JsonUnquoteFunction.java:78)
	at org.apache.flink.table.runtime.functions.scalar.JsonUnquoteFunction.eval(JsonUnquoteFunction.java:112)
	... 11 more
```

UPD: i think it is better to follow existing's vendor's way like MySQL and do not allow more than 2 non-escape double quotes

fhueske reviewed (commented):
Thanks for working on this @anupamaggarwal!

I have just a few high-level comments so far.
Will dig a bit deeper into this soon.

Thank you, Fabian

anupamaggarwal commented:
thanks @snuyanzin for your review!
I think both these cases are related to inconsistent behavior (IIUC) of `IS JSON` function which is used in json_unquote implementation for checking JSON validity. I had observed something similar earlier (see `Open questions` section of the PR description. 

```
SELECT JSON_UNQUOTE('"1""1"');
```
As you pointed out this should ideally throw an exception (as mysql), however IS JSON returns true in this case  (IIUC this should be a bug? since `1""1` is not a valid json)

```
select '"1""1"' is json; //true 
```  
If we fix this by throwing an exception (preferable), it could be in-consistent with what IS JSON considers as valid json?
Should we care about compatibility with IS JSON behavior here (or just fix this and file a JIRA for IS JSON?)

Same is the case with 
```
SELECT JSON_UNQUOTE('"1""\uzzzz"')
select '"1""\uzzzz"' is JSON; //true
``` 
However in this case we throw an exception (from json_unquote) since \uzzzz is not a valid literal 

For 
```
SELECT JSON_UNQUOTE('"1\uzzzz"')
```
 IS JSON returns FALSE and we return the input string back per documentation. 

>SELECT JSON_UNQUOTE('"\"\ufffa"');
>it starts printing backslash as well which is NOT OK

Hmm, I tested this through the sql-client.sh, and  I get `" ` which seems expected? (as "\"\ufffa" is a valid JSON )

anupamaggarwal commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala:
```diff
@@ -604,6 +604,49 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
+  }
+
+  @Test
+  def testJsonUnquote(): Unit = {
```
thanks @fhueske for your review, will fix this

snuyanzin commented:
Thanks for the explanation
>However in this case we throw an exception (from json_unquote) since \uzzzz is not a valid literal

that's the main point, since in related docs of this PR is written [1]
>If the value starts and ends with double quotes but is not a valid JSON string literal, the value is passed through unmodified.

So based on the doc instead of exception I would expect same value as it was in input

[1] https://github.com/apache/flink/pull/24967/files#diff-539fb22ee6aeee4cf07230bb4155500c6680c4cc889260e2c58bfa9d63fb7de5R385

fhueske commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/JsonUnquoteFunction.java:
```diff
@@ -0,0 +1,118 @@
+    private String unquote(String inputStr) {
+
+        StringBuilder result = new StringBuilder();
+        int i = 1;
```
we are skipping the first and the last character without checking whether they are double quotes.
The string `[1, 2, 3]` is valid JSON and we return `1, 2, 3`.
I think we also need to check if the input string is surrounded by double quotes (even before checking if it's valid JSON) and return the input string if the quotes are missing.

For reference, this is MySQL's output:
```
mysql> select json_unquote('[1, 2, 3]');
+---------------------------+
| json_unquote('[1, 2, 3]') |
+---------------------------+
| [1, 2, 3]                 |
+---------------------------+
1 row in set (0.00 sec)

mysql> select json_unquote('"[1, 2, 3]"');
+-----------------------------+
| json_unquote('"[1, 2, 3]"') |
+-----------------------------+
| [1, 2, 3]                   |
+-----------------------------+
1 row in set (0.00 sec)
```

These cases should also be covered by tests.

anupamaggarwal replied:
Great catch @fhueske 👍, I completely missed handling this.

fhueske reviewed (changes requested):
Thanks for the update @anupamaggarwal!

I've found another case that we are not handling correctly.
Luckily, it doesn't depend on `IS_JSON` and easy to fix :-D

Thanks, Fabian

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/JsonUnquoteFunction.java:
```diff
@@ -0,0 +1,128 @@
+        return String.valueOf((char) code);
+    }
+
+    public @Nullable Object eval(Object input) {
```
nit: usually we put public methods above and utility private methods below
e.g. https://github.com/apache/flink/blob/master/flink-table/flink-sql-parser/src/main/java/org/apache/calcite/sql/fun/SqlJsonQueryFunction.java

fhueske commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/JsonUnquoteFunction.java:
```diff
@@ -0,0 +1,128 @@
+                        i = i + 4;
+                        break;
+                    default:
+                        throw new RuntimeException("Illegal escape sequence: \\" + ch);
```
```suggestion
                        throw new IllegalArgumentException("Illegal escape sequence: \\" + ch);
```

snuyanzin replied:
IMHO: since it is already runtime shouldn't it be `FlinkRuntimeException` here?
https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/util/FlinkRuntimeException.java

which is already used within builtin functions at https://github.com/apache/flink/tree/master/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar

fhueske commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/JsonUnquoteFunction.java:
```diff
@@ -0,0 +1,128 @@
+            if (isValidJsonVal(inputStr)) {
+                return new BinaryStringData(unescapeValidJson(inputStr));
+            }
+        } catch (Throwable t) {
```
I don't think we should swallow any kind of exception or error.
Let's catch the errors that we expect (`IllegalArgumentException`, etc.) and handle those by returning the original input.
Unexpected exceptions should be forwarded because they might be related to other issues.

fhueske replied:
sorry, commented on an outdated version...

fhueske reviewed (commented):
Thanks for adding the quote checks and additional test cases.
I found one more thing that we should fix.

Thank you, Fabian

fhueske commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/JsonUnquoteFunction.java:
```diff
@@ -0,0 +1,128 @@
+            if (isValidJsonVal(inputStr)) {
+                return new BinaryStringData(unescapeValidJson(inputStr));
+            }
+        } catch (Throwable t) {
+            // ignore
+        }
```
I don't think we should swallow any kind of exception or error.
Let's catch the errors that we expect (IllegalArgumentException, etc.) and handle those by returning the original input.
Unexpected exceptions should be forwarded because they might be related to other issues.

anupamaggarwal replied:
thanks @fhueske, @snuyanzin  makes sense, fixed

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/JsonUnquoteFunction.java:
```diff
@@ -0,0 +1,129 @@
+            if (isValidJsonVal(inputStr)) {
+                return new BinaryStringData(unescapeValidJson(inputStr));
+            }
+        } catch (FlinkRuntimeException | IllegalArgumentException t) {
```
Ooops, seems I was wrong about `FlinkRuntimeException`
if we want to rethrow it further then it would make sense

however of we use it only within this class then `IllegalArgumentException` is enough
sorry about that
I should have checked the whole cycle of it here

snuyanzin commented:
One more diff between MySQL and Flink behavior, however this I'm not sure need to fix since at least from my point of view MySQL's behavior is questionable
MySQL
```sql
SELECT  JSON_UNQUOTE('"\\u0022\\u005c\\u005c\\u0075\\u0030\\u0030\\u0061\\u0061\\u0022"');
SELECT  JSON_UNQUOTE(JSON_UNQUOTE('"\\u0022\\u005c\\u005c\\u0075\\u0030\\u0030\\u0061\\u0061\\u0022"'));
```
returns
```
"\\u00aa"
\u00aa  
```
Flink
```sql
SELECT  JSON_UNQUOTE('"\u0022\u005c\u005c\u0075\u0030\u0030\u0061\u0061\u0022"');
SELECT  JSON_UNQUOTE(JSON_UNQUOTE('"\u0022\u005c\u005c\u0075\u0030\u0030\u0061\u0061\u0022"'));
```
returns 
```
"\\u00aa"
ª 
```

anupamaggarwal commented:
thanks @snuyanzin, fixed in [7546d9c](https://github.com/apache/flink/pull/24967/commits/7546d9ceb85b5650804d6868bb34797507076c16)

Hmm, that's strange, when I tested 
>SELECT  JSON_UNQUOTE('"\u0022\u005c\u005c\u0075\u0030\u0030\u0061\u0061\u0022"');
SELECT  JSON_UNQUOTE(JSON_UNQUOTE('"\u0022\u005c\u005c\u0075\u0030\u0030\u0061\u0061\u0022"'));
 
 through sql-client I get the same result as MySql
 
 I didn't entirely understand why this would be unexpected though
For example, if I test (using sql-client) with
 
 `select json_unquote('"\u00aa"');` I get `ª`
 whereas 
 `select json_unquote('"\\u00aa"');` returns `\u00aa`

 which seems ok to me, unless I missed something?
 
 Added some tests in `ScalarFunctionsTest`

fhueske commented on docs/data/sql_functions_zh.yml:
```diff
@@ -477,6 +477,12 @@ string:
+    description: Quotes a string as a JSON value by wrapping it with double quote characters, escaping interior quote and special characters ('"', '\', '/', 'b', 'f', 'n', 'r', 't'), and returning the result as a string. If the argument is NULL, the function returns NULL.
+  - sql: JSON_UNQUOTE(string)
+    table: STRING.JsonUnquote()
+    description: Unquotes JSON value, unescapes escaped special characters ('"', '\', '/', 'b', 'f', 'n', 'r', 't', 'u' hex hex hex hex), and returns the result as a string. If the argument is NULL, returns NULL. If the value starts and ends with double quotes but is not a valid JSON string literal, the value is passed through unmodified.
```
```suggestion
    description: Unquotes JSON value, unescapes escaped special characters ('"', '\', '/', 'b', 'f', 'n', 'r', 't', 'u' hex hex hex hex), and returns the result as a string. If the argument is NULL, returns NULL. If the value does not start and ends with double quotes or if it starts and ends with double quotes but is not a valid JSON string literal, the value is passed through unmodified.
```

anupamaggarwal replied:
Fixed in https://github.com/apache/flink/pull/24967/commits/37fa2e78cb88249eb448e01445a5c3ce9197608f

fhueske reviewed (approved):
Thanks for the update @anupamaggarwal.

I've double checked the behavior of 
> SELECT JSON_UNQUOTE('"\u0022\u005c\u005c\u0075\u0030\u0030\u0061\u0061\u0022"');
SELECT JSON_UNQUOTE(JSON_UNQUOTE('"\u0022\u005c\u005c\u0075\u0030\u0030\u0061\u0061\u0022"'));

and can confirm that it is the same as in MySQL.

I think this PR is ready to be merged (just one improvement of the docs).
What do you think @snuyanzin?

snuyanzin commented:
yes, probably there is something wrong with my mysql...
thanks for checking
and thanks for addressing feedback

snuyanzin commented:
@anupamaggarwal  can you please rebase to the latest master to be sure that it still passes ci ? (I tend to think that current failure is not related to the changes in PR)

anupamaggarwal commented:
@flinkbot run azure

anupamaggarwal commented:
Thanks @snuyanzin I merged the latest from master and the build is passing now (It seems CI is a bit flaky since one of the runs post merge failed but the second went through). The failure didn't seem related to this change though.

snuyanzin commented:
Cool, thanks for your contribution @anupamaggarwal 
thanks for the review @fhueske 

if there is no objections I'm going to merge later today or tomorrow

fhueske commented:
Thank you both @anupamaggarwal and @snuyanzin!
